### PR TITLE
GeoFit FSR update Grouped JEC for 2016

### PR DESCRIPTION
**Pull request by nanlu06**
changes:

1. Adding GeoFit:

muon pt after correction is calculated on the fly. The macro here
https://github.com/UFLX2MuMu/H2MuAnalyzer/blob/master/GeoFit/GeoFitCorr.C

d0_BS used by UF group: muon.innerTrack()->dxy( beamSpotHandle->position() )

while nanoAOD uses best track

2. FSR update:

Update the FSR recovery to Oliver's slide 8 slide 8
https://indico.cern.ch/event/867116/contributions/3654090/subcontributions/294675/attachments/1957523/3252192/h2mu_workingmeetin05122019.pdf

3. Reduced JEC for 2016

4. new samples for 2016: data/datasets_NanoAODv6_2016_mixv1.yml

### CODE REVIEW

**Comment by jpata** [comment]
Thanks Nan! I noticed the automatic test is failing with `KeyError: not found: b'Muon_dxybs'`. We could upload a new test file from the private production, to make sure that the code always has some basic sanity checks.

**Comment by nanlu06** [comment]
Thank you, Joosep! Would it work to upload new files to here?

data/nano_2016_data.root

data/myNanoProdMc2016_NANO.root

https://github.com/jpata/hepaccelerate-cms/blob/c8255e27268eea186f62dd113f263208ad37bac9/tests/hmm/test_analysis_full.py#L44

https://github.com/jpata/hepaccelerate-cms/blob/c8255e27268eea186f62dd113f263208ad37bac9/tests/hmm/test_analysis_full.py#L48

**Comment by nanlu06** [comment]
datacards and histograms: https://cernbox.cern.ch/index.php/s/J2CYzXo4i7HvJyt. 

Looks OK.

File: tests/hmm/hmumu_utils.py [hunk 1]
```diff
@@ -1144,17 +1161,33 @@ def fix_muon_fsrphoton_index(offsets_fsrphotons, offsets_muons, fsrphotons_dROve
             #Index of FSR photon associated to muon
             fidx = offsets_fsrphotons[iev] + muons_fsrPhotonIdx[midx]
 
-            if k != muons_fsrPhotonIdx[midx]:
-                if fsrphotons_dROverEt2[i] < fsrphotons_dROverEt2[fidx]:
-                    out_muons_fsrPhotonIdx[midx] = k
+            px = mu_pt[midx] * np.cos(mu_phi[midx])
```
Comment (jpata) [inline tests/hmm/hmumu_utils.py]: could also use`hepaccelerate.spherical_to_cartesian` here: https://github.com/hepaccelerate/hepaccelerate/blob/master/hepaccelerate/kernels.py#L4

File: tests/hmm/hmumu_utils.py [hunk 2]
```diff
@@ -2026,8 +2063,12 @@ def get_gen_sumweights(filenames):
     for fi in filenames:
         ff = uproot.open(fi)
         bl = ff.get("Runs")
-        arr = bl.array("genEventSumw")
-        arr2 = bl.array("genEventSumw2")
+        try:
+            arr = bl.array("genEventSumw")
+            arr2 = bl.array("genEventSumw2")
+        except KeyError:
+            arr = bl.array("genEventSumw_")
```
Comment (jpata) [inline tests/hmm/hmumu_utils.py]: might be good to put a comment here for future reference, perhaps also a printout.

**Comment by irenedutta23** [comment]
So just for my understanding, I see there are some JECS that have an overall component and also a per year component (for eg. the BBEC1 and BBEC2016/17/18 ). Is it something like the BBEC2016 would be just a flat distribution for 2017 samples? and the overall component takes into account the correlation amongst different years?

**Comment by nanlu06** [comment]
Hi Irene! Yes, uncertainties like BBEC1 are 100% correlated across years, and the other uncertainties like BBEC2016/17/18 are separated for different years, so totally uncorrelated across years. BBEC2016 would be the same distribution as the nominal histogram for 2017 samples, as the way is implemented in our analysis code.

File: tests/hmm/hmumu_utils.py (hunk 2 above)
Comment (nanlu06) [reply tests/hmm/hmumu_utils.py]: added.

File: tests/hmm/hmumu_utils.py (hunk 1 above)
Comment (nanlu06) [reply tests/hmm/hmumu_utils.py]: Thank you for pointing out this info! will improve in the next PR.

**Pull request merged**